### PR TITLE
Errors on non-immediate load/gather offsets

### LINKED ISSUE #2590: Texture.Load with non constant offsets
Hi,

Looks like DXC does not complain when using non compile-time constant offsets with Texture.Load, but my understanding is that it should?

eg. http://shader-playground.timjones.io/7a66a2a01b6e6522fce6fef5ed312678

For equivalent code, FXC reports either
`error X3527: texture access must have literal offset and multisample index` or
`error X4566: offset texture instructions must take offset which can resolve to integer literal in the range -8 to 7`

It is kind of related to https://github.com/microsoft/DirectXShaderCompiler/issues/2562 I suppose, I had the same issue with code that could be valid if the loop was unrolled, but DXC didn't unroll it, and then it crashed in the driver during the PSO creation on AMD (with DX12).

### LINKED ISSUE #2713: Missing validation of integer offsets for Gather
Hi,

I think FXC and DXC are both missing validation against using non-literal offsets and also offsets that are literal but outside the required range of -8 to +7. The D3D11.3 spec* confirms these restrictions on Sample and Load operations but makes no mention of them existing on Gather.

However, based on the way GCN hardware appears to pack its offsets (into no more than a byte each) there’s no way they can support offsets of more than about -128 - +127 (and in practice it seems to be somewhere around -64 to +63.

[Shader](http://shader-playground.timjones.io/05cef670521469e423bb10db6599f0c4)

main3 (Sample with big offsets) and main4 (Sample with offsets that aren’t compile time literals) give this error and I believe main and main2 probably should too:

G:\gather.txt(7,12-39): error X4566: offset texture instructions must take offset which can resolve to integer literal in the range -8 to 7

The D3D11.3 spec may also want updating if this is an omission.

Thanks,

Adam

*
Address Offset
--------------
The optional [_aoffimmi(u,v,w)] suffix (address offset by
immediate integer) indicates that the texture coordinates for
the sample are to be offset by a set of provided immediate
texel space integer constant values. The literal values are a
set of 4 bit 2's complement numbers, having integer range
 [-8,7]. This modifier is defined for all Resources,
including Texture1D/2D Arrays and Texture3D,
but it is undefined for TextureCube.

### PR DESCRIPTION
**Pull request by pow2clk**
Add errors for gather and load intrinsics when offsets are non-
literals or out of range. These largely use the slightly altered
validation paths that the sample intrinsics use.

Reword error when texture access offsets are not immediates to be
more all encompassing and grammatical.

Incendentally remove duplicate shaders being used for the validation
test from the old directory while identical copies in the validation
directory went unused. Redirected validation test to the appropriate
copies. This is consistent with the test shader re-org's stated intent

Fixes #2590
Fixes #2713

### CODE REVIEW

File: tools/clang/test/HLSLFileCheck/hlsl/objects/Texture/gatherOffset.hlsl [hunk 1]
```diff
@@ -35,9 +35,9 @@ float4 main(float4 a : A, float4 b : B) : SV_Target
 {
   uint status;
   float4 r = 0;
-  r += text1.GatherRed(samp1, a.xy, b.xy, b.zw, a.xy, a.zw);
-  r += text1.GatherAlpha(samp1, a.xy, b.xy, b.zw, a.xy, a.zw, status); r += status;
+  r += text1.GatherRed(samp1, a.xy, int2(0, 0), int2(-1, 1), int2(3, 4), int2(-8, 7));
+  r += text1.GatherAlpha(samp1, a.xy, int2(0, 0), int2(-1, 1), int2(3, 4), int2(-8, 7), status); r += status;
 
-  r += text2.GatherBlue(samp1, a.xyz, b.xy, b.zw, a.xy, a.zw, status); r += status;
+  r += text2.GatherBlue(samp1, a.xyz, int2(0, 0), int2(-1, 1), int2(3, 4), int2(-8, 7), status); r += status;
```
Comment (pow2clk) [inline tools/clang/test/HLSLFileCheck/hlsl/objects/Texture/gatherOffset.hlsl:41]: These old tests flagrantly violating this purported restriction and the lack of any documentation of it do make me wonder if maybe this should be allowed. When next I have access to FXC, I'll see what that does with these.

Comment (pow2clk) [reply tools/clang/test/HLSLFileCheck/hlsl/objects/Texture/gatherOffset.hlsl:41]: I still think this is correct, but for the record, FXC produces errors on out of range and variable offset parameters for Load and Samples, but not for Gathers.

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp [hunk 2]
```diff
@@ -182,6 +181,9 @@ void DxilLegalizeSampleOffsetPass::CollectIllegalOffsets(
   CollectIllegalOffsets(illegalOffsets, CurF, DXIL::OpCode::SampleGrad, hlslOP);
   CollectIllegalOffsets(illegalOffsets, CurF, DXIL::OpCode::SampleLevel,
                         hlslOP);
+  CollectIllegalOffsets(illegalOffsets, CurF, DXIL::OpCode::TextureGather, hlslOP);
+  CollectIllegalOffsets(illegalOffsets, CurF, DXIL::OpCode::TextureGatherCmp, hlslOP);
+  CollectIllegalOffsets(illegalOffsets, CurF, DXIL::OpCode::TextureLoad, hlslOP);
```
Comment (vcsharma) [inline lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:186]: nit: given that we are now expanding scope of this pass to also legalize offsets for texture gather and load ops, it might be more clearer to rename this pass as `DxilLegalizeOffsetsPass` or something else which more accurately reflects the expanded scope.

File: lib/HLSL/DxilValidation.cpp [hunk 3]
```diff
@@ -1261,7 +1261,7 @@ static void ValidateResourceOffset(CallInst *CI, DXIL::ResourceKind resKind,
           validateOffset(offsets[i]);
       }
     } else {
-      if (!isa<UndefValue>(offsets[i])) {
+      if (i < offsets.size() && !isa<UndefValue>(offsets[i])) {
```
Comment (vcsharma) [inline lib/HLSL/DxilValidation.cpp:1264]: It's not very clear to me why we need to do additional bound check here (`i < offsets.size()`). Is it because we can no longer expect `kMaxNumOffsets` to be 3? It might change based on resource kind?

File: tools/clang/test/HLSLFileCheck/hlsl/diagnostics/errors/offsets.hlsl [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+// CHK_RANGE:  error: offset texture instructions must take offset which can resolve to integer literal in the range -8 to 7.
+
+// CHK_VAROFF: Offsets to texture access operations must be immediate values
+// CHK_VAROFF: Offsets to texture access operations must be immediate values
+// CHK_VAROFF: Offsets to texture access operations must be immediate values
+// CHK_VAROFF: Offsets to texture access operations must be immediate values
+// CHK_VAROFF: Offsets to texture access operations must be immediate values
+// CHK_VAROFF: Offsets to texture access operations must be immediate values
+// CHK_VAROFF: Offsets to texture access operations must be immediate values
+// CHK_VAROFF: Offsets to texture access operations must be immediate values
```
Comment (vcsharma) [inline tools/clang/test/HLSLFileCheck/hlsl/diagnostics/errors/offsets.hlsl:16]: Nice!

**Review by vcsharma — COMMENTED** [review]
This change looks good, but it may also unearth existing shaders/tests using offsets which are not immediate for texture load/gather ops. As an example, I shared with you offline a failing internal test that we might want to address it as part of this change.

File: lib/HLSL/DxilValidation.cpp (hunk 3 above)
Comment (pow2clk) [reply lib/HLSL/DxilValidation.cpp:1264]: It's been so long I don't remember exactly what I was thinking here. You are correct that `kMaxNumOffsets` should no longer be 3 in all cases. It could have needed to be 4 for the existing cases before too. It shouldn't really be a constant either.

I think I'll remove this conditional and change the `kMaxNumOffsets` loop to use the size of `offsets` and add some more tests for cases that have different expected numbers of offsets and different maximum numbers of offsets

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp (hunk 2 above)
Comment (tex3d) [reply lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:186]: Perhaps, but in my opinion, the name is ok, as it encompasses offsets to texture sample locations for textures, and not offsets for a variety of other things we might use the word "offset" for, even if it's for more than just the "Sample*()" methods.

File: lib/HLSL/DxilValidation.cpp (hunk 3 above)
Comment (tex3d) [reply lib/HLSL/DxilValidation.cpp:1264]: Futher up in the stack, only the expected offsets are passed into this function as an array ref.  This array size is not always kMaxNumOffsets, so this bound check prevents OOB index on offsets[i].

Comment (tex3d) [reply lib/HLSL/DxilValidation.cpp:1264]: @pow2clk  I think that should be ok: using `offsets.size()` instead of `kMaxNumOffsets` for loop, and removing the bound check here.  I think this was leftover from a version of the code that had an assumed fixed sized array.

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp [hunk 5]
```diff
@@ -112,16 +112,15 @@ bool HasIllegalOffsetInLoop(std::vector<Instruction *> &illegalOffsets,
...
+      CI->getArgOperand(offsetIdx);
   // No offset.
   if (isa<UndefValue>(offset0))
     return;
 
-  for (unsigned i = DXIL::OperandIndex::kTextureSampleOffset0OpIdx;
-       i <= DXIL::OperandIndex::kTextureSampleOffset2OpIdx; i++) {
+  for (unsigned i = offsetIdx; i <= offsetIdx + 2; i++) {
     Value *offset = CI->getArgOperand(i);
     if (Instruction *I = dyn_cast<Instruction>(offset))
```
Comment (tex3d) [inline lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:125]: Shouldn't we be checking for constant offsets out of `[-8..7]` range here as well, and collecting these to a separate list (intrinsic instruction instead of argument, since argument would be constant)?

Comment (pow2clk) [reply lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:125]: The intent of this pass is to try to make the samples constant if possible. the generation of an error is incidental. I could tack on a check for the ranges here if it would be of benefit, but I'm not sure it would be much different from the validation location.

In fact, the followup change I'm working on has had to revise some of this code more than I originally intended to get errors that are of any use since the existing code only keeps around the offset source, which is often not something with a debug location and not as helpful as the calling function that takes the offsets anyway.

**Comment by pow2clk** [comment]
I divided the new changes into logically independent commits. It might be easier to review them individually.

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp [hunk 6]
```diff
@@ -198,19 +260,19 @@ void DxilLegalizeSampleOffsetPass::CollectIllegalOffsets(
       if (CI->getParent()->getParent() != &CurF)
         continue;
 
-      CollectIllegalOffset(CI, illegalOffsets);
+      CollectIllegalOffset(CI, opcode, illegalOffsets);
```
Comment (pow2clk) [inline lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:263]: opcode included to determine offset operand locations. It could be derived from the CI, but it's already available here anyway.

File: lib/HLSL/DxilPreparePasses.cpp [hunk 7]
```diff
@@ -381,6 +381,29 @@ class DxilFinalizeModule : public ModulePass {
...
+        for (unsigned i = 1; i < CI->getNumArgOperands(); i++)
+          args.emplace_back(CI->getOperand(i));
+
+        Function *newF = hlslOp->GetOpFunc(toOpcode, Ty);
+        CallInst *NewCI = Builder.CreateCall(newF, args);
+        CI->replaceAllUsesWith(NewCI);
+        CI->eraseFromParent();
+      }
+    }
+  }
```
Comment (pow2clk) [inline lib/HLSL/DxilPreparePasses.cpp:405]: Generic only because there was no reason not to make it so. Used exclusively to replace immediate restricted gathers with their established alternatives until we add the immediate versions to the DXIL spec

File: lib/HLSL/DxilValidation.cpp [hunk 8]
```diff
@@ -1390,8 +1389,11 @@ static void ValidateGather(CallInst *CI, Value *srvHandle, Value *samplerHandle,
   default:
     // Invalid resource type for gather.
     ValCtx.EmitInstrError(CI, ValidationRule::InstrResourceKindForGather);
-    break;
+    return;
```
Comment (pow2clk) [inline lib/HLSL/DxilValidation.cpp:1392]: Return on error to prevent calling ValidateResourceOffset on an invalid resource as happens with some of our tests specifically targeting the validator by altering DXIL after it is produced

File: lib/HLSL/DxilValidation.cpp [hunk 9]
```diff
@@ -2199,8 +2203,11 @@ static void ValidateResourceDxilOp(CallInst *CI, DXIL::OpCode opcode,
     default:
       ValCtx.EmitInstrError(CI,
                             ValidationRule::InstrResourceKindForTextureLoad);
-      break;
+      return;
```
Comment (pow2clk) [inline lib/HLSL/DxilValidation.cpp:2206]: As above, return prevents attempts to validate invalid resources as an error is already produced

File: lib/HLSL/HLOperationLower.cpp [hunk 10]
```diff
@@ -3167,9 +3167,12 @@ GatherHelper::GatherHelper(
       if (ch != GatherChannel::GatherAll)
         TranslateSampleOffset(CI, HLOperandIndex::kGatherSampleOffsetArgIndex,
                               offsetSize);
-      statusIdx =
-          hasSampleOffsets ? HLOperandIndex::kGatherStatusWithSampleOffsetArgIndex
-                           : HLOperandIndex::kGatherStatusArgIndex;
+      if (hasSampleOffsets) {
+        statusIdx = HLOperandIndex::kGatherStatusWithSampleOffsetArgIndex;
+      } else {
+        opcode = OP::OpCode::TextureGatherImm;
```
Comment (pow2clk) [inline lib/HLSL/HLOperationLower.cpp:3173]: Where there aren't 4 offset parameters, the immediate restricted gather should be used.

File: tools/clang/test/CodeGenHLSL/optForNoOpt3.hlsl [hunk 11]
```diff
@@ -1,17 +0,0 @@
-// RUN: %dxc -E main -T ps_6_0 -Od %s | FileCheck %s
```
Comment (pow2clk) [inline tools/clang/test/CodeGenHLSL/optForNoOpt3.hlsl:1]: These files were redundant. Where they were used, the more appropriate location was in dxilvalidation/ where they were copied with the mass test file migration probably with the intent to make this exact change. Rather than alter these as they should be, I resolved what seemed like a mistake.

File: tools/clang/test/HLSLFileCheck/hlsl/objects/Texture/gatherCmp.hlsl [hunk 12]
```diff
@@ -14,26 +14,26 @@ float4 main(float2 a : A) : SV_Target
   r += text1.GatherCmp(samp1, a, cmpVal, uint2(-3, 2), status); r += status;
 
   r += text1.GatherCmpAlpha(samp1, a, cmpVal);
-  r += text1.GatherCmpAlpha(samp1, a, cmpVal, uint2(-3,8));  
-  r += text1.GatherCmpAlpha(samp1, a, cmpVal, uint2(-3,8),status); r += status;    
+  r += text1.GatherCmpAlpha(samp1, a, cmpVal, uint2(-3,7));
+  r += text1.GatherCmpAlpha(samp1, a, cmpVal, uint2(-3,7),status); r += status;
   r += text1.GatherCmpAlpha(samp1, a, cmpVal, uint2(-3,8),uint2(-2,3), uint2(-3,8),uint2(-2,3));  
   r += text1.GatherCmpAlpha(samp1, a, cmpVal, uint2(-3,8),uint2(8,-3), uint2(8,-3), uint2(-3,2), status); r+=status;
```
Comment (pow2clk) [inline tools/clang/test/HLSLFileCheck/hlsl/objects/Texture/gatherCmp.hlsl:20]: Chose to leave the 4-offset variant with values outside the range even though I doubt that was the original intent. It verifies that these are still possible

**Review by pow2clk — COMMENTED** [review]
Explanatory comments

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp [hunk 13]
```diff
@@ -68,94 +70,151 @@ class DxilLegalizeSampleOffsetPass : public FunctionPass {
...
 
-  for (unsigned i = DXIL::OperandIndex::kTextureSampleOffset0OpIdx;
-       i <= DXIL::OperandIndex::kTextureSampleOffset2OpIdx; i++) {
+  for (unsigned i = offsetStart; i <= offsetEnd; i++) {
     Value *offset = CI->getArgOperand(i);
     if (Instruction *I = dyn_cast<Instruction>(offset))
-      illegalOffsets.emplace_back(I);
+      illegalOffsets.emplace_back(std::make_pair(I, CI));
+    else if(ConstantInt *cOffset = dyn_cast<ConstantInt>(offset)) {
+      int val = cOffset->getValue().getSExtValue();
```
Comment (tex3d) [inline lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:157]: This returns `int64_t`, and you are relying on implicit cast to int, recommend making this an explicit cast so we don't get build breaks.

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp [hunk 14]
```diff
@@ -68,94 +70,151 @@ class DxilLegalizeSampleOffsetPass : public FunctionPass {
...
+
+// Return true if the call instruction in pair a is before that in pair b
+bool InstLT(const std::pair<Value*,CallInst*> &a, const std::pair<Value*,CallInst*> &b) {
+  DebugLoc aLoc = a.second->getDebugLoc();
+  DebugLoc bLoc = b.second->getDebugLoc();
+  if (aLoc && bLoc)
+    return aLoc.getLine() < bLoc.getLine();
+  // No line numbers, just compare pointers so that matching instructions will be adjacent
+  return a.second < b.second;
+}
```
[the remark this answers is not in the document]
Comment (tex3d) [reply lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:178]: Nice, but what about source file?  Won't this interleave messages based on line number from different source files?

Additionally, I don't think pointer compares will produce deterministic output.  Even though that's in the error message ordering and not generated code, it still seems yucky.

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp [hunk 15]
```diff
@@ -68,94 +70,151 @@ class DxilLegalizeSampleOffsetPass : public FunctionPass {
...
-    raw_string_ostream errorStr(errorMsg);
-    for (Instruction *offset : finalIllegalOffsets)
-      dxilutil::EmitErrorOnInstruction(offset, kIllegalOffsetError);
+    std::string errorMsg = "Offsets to texture access operations must be immediate values. ";
+
+    auto offsetBegin = finalIllegalOffsets.begin();
+    auto offsetEnd = finalIllegalOffsets.end();
+
+    std::sort(offsetBegin, offsetEnd, InstLT);
+    offsetEnd = std::unique(offsetBegin, offsetEnd, InstEq);
```
[the remark this answers is not in the document]
Comment (tex3d) [reply lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:192]: Maybe we could avoid re-ordering based on debug info by checking for debug info first?  Then reordering by pointers doesn't have to happen and the order can at least be deterministic?

**Review by tex3d — CHANGES_REQUESTED** [review]

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp [hunk 16]
```diff
@@ -57,7 +59,7 @@ class DxilLegalizeSampleOffsetPass : public FunctionPass {
     DxilModule &DM = F.getParent()->GetOrCreateDxilModule();
     hlsl::OP *hlslOP = DM.GetOP();
 
-    std::vector<Instruction *> illegalOffsets;
+    std::vector<std::pair<Value *, CallInst*>> illegalOffsets;
```
[the remark this answers is not in the document]
Comment (pow2clk) [reply lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:62]: If I'm going to have a name, I might as well make a struct that allows meaningful field access

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp [hunk 17]
```diff
@@ -68,94 +70,151 @@ class DxilLegalizeSampleOffsetPass : public FunctionPass {
...
-  for (Instruction *I : illegalOffsets) {
-    BasicBlock *BB = I->getParent();
-    if (LI.getLoopFor(BB)) {
-      findOffset = true;
-      break;
+  for (auto it : illegalOffsets) {
+    if (Instruction *I = dyn_cast<Instruction>(it.first)) {
+      BasicBlock *BB = I->getParent();
+      if (LI.getLoopFor(BB)) {
+        findOffset = true;
```
[the remark this answers is not in the document]
Comment (pow2clk) [reply lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:109]: I didn't actually change any of the function of this code. @adam-yang might have thoughts on that suggestion. 

Nevertheless, I'm happy to add the TODO.

The error message makes clear that manual unrolling is a speculative fix.

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp (hunk 13 above)
Comment (pow2clk) [reply lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:157]: No reason not to make val int64_t

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (pow2clk) [reply lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:178]: pointer compares are not deterministic across compiles and the results are weird, but without debug info, we don't get any line numbers anyway so it's just the same error repeated. It does allow elimination of duplicates within a single compile which means at least the number of errors is right

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (pow2clk) [reply lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:192]: As I said above, ordering by pointers is weird, but it at least allows eliminating duplicates

**Review by pow2clk — COMMENTED** [review]
Thanks @tex3d!

Responses inline

File: lib/HLSL/DxilLegalizeSampleOffsetPass.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (tex3d) [reply lib/HLSL/DxilLegalizeSampleOffsetPass.cpp:119]: Yes, this was just something I thought of while reading the code that could be captured somewhere relevant, like in a TODO comment, rather than just forgetting about it again.  Of course, it only makes sense to add a comment when already updating the PR for a better reason (and there were other reasons to do so).

**Review by tex3d — APPROVED** [review]
